src/algorithm/solarPanelPlacement.ts

```diff
@@ -0,0 +1,259 @@
+
+  for (let i = 0; i < matrixLength; i++) {
+    for (let j = 0; j < matrixWidth; j++) {
+      if (matrix[i][j].usable) {
+        const solarPanel = createSolarPanel([i, j], horizontal)
+        if (isAllSolarPanelGridUsable(matrix, solarPanel)) {
+          if (isAllSolarPanelGridNotCovered(solarPanel, coverageMatrix)) {
+            solarPanels.push(solarPanel)
+            for (let k = 0; k < solarPanel.gridIndexes.length; k++) {
+              const x = solarPanel.gridIndexes[k][0]
+              const y = solarPanel.gridIndexes[k][1]
+              // Set coverage matrix to the last index of the solar panel
+              coverageMatrix[x][y] = solarPanels.length - 1
+            }
+          } else {
+            // skipped because the solar panel cover grid that has been covered by other solar panel
+          }
+        } else {
+          // skipped because the grids covered by the solar panel are not all usable or beyond the matrix
+        }
+      } else {
+        // skipped if the grid is not usable
+      }
+    }
+  }
+
+  return solarPanels
+}
+
+export function solarPanelPlacement(
+  matrix: Matrix,
+  debug: boolean = false,
+  prefix: string = ''
+) {
+  const matrixGeoJSON = matrixToGeoJSON(matrix)
+  writeFeature(`${prefix}matrix.geojson`, matrixGeoJSON, debug)
+
+  const centroidGeoJSON = matrixCentroidToGeoJSON(matrix)
+  writeFeature(`${prefix}centroids.geojson`, centroidGeoJSON, debug)
+
+  const verticalSolarPanels = solarPanelPlacementAlgorithm(matrix, false)
+  const horizontalSolarPanels = solarPanelPlacementAlgorithm(matrix, true)
+
+  let orientation
+  let solarPanels
+  if (horizontalSolarPanels.length >= verticalSolarPanels.length) {
+    solarPanels = horizontalSolarPanels
+    orientation = 'horizontal'
+  } else {
+    solarPanels = verticalSolarPanels
+    orientation = 'vertical'
+  }
+
+  const matrixGeoJSONCovered = matrixToGeoJSON(matrix, solarPanels)
+  writeFeature(`${prefix}matrix-covered.geojson`, matrixGeoJSONCovered, debug)
+
+  return { solarPanels: solarPanels, orientation: orientation }
+}
+
+// Used to write a GeoJSON representation of the matrix' to check
+// the order / formation of the grid and the solar panel
+export function matrixToGeoJSON(
+  matrix: Matrix,
+  solarPanels: SolarPanelGrid[] = []
+) {
+  const fc: FeatureCollection<Polygon, Properties> = featureCollection([])
+  const length = 0.05
+
+  const coveredGrids: Record<number, number> = {}
+  for (let i = 0; i < solarPanels.length; i++) {
+    solarPanels[i].gridIndexes.forEach((gridIndex) => {
+      coveredGrids[gridIndex[1] * matrix.length + gridIndex[0]] = i
+    })
+  }
+
+  for (let i = 0; i < matrix.length; i++) {
+    for (let j = 0; j < matrix[0].length; j++) {
+      const x: number = i * -length
+      const y: number = j * length
+      const solarPanelIndex = coveredGrids[j * matrix.length + i]
+      const grid: Feature<Polygon, Properties> = polygon(
```

Comment:
Can you explain here in comment please, It's a pretty complicated part to understand.